### ISSUE TITLE
Query throw divide/zero exception with Decimal

### ISSUE THREAD

**Issue by 5herhom** [ClickHouse/ClickHouse#24555]
According to the 2 pictures below,  the p4_1y_comment_l_score and p4_1y_comment_h_score are all type decimal(32,6).customer_no is string type.
It is running in version 20.6.3.28。
When use customer_no to projection , it throw divide/zero exception.
sql1:
select customer_no from xxxx where p4_1y_comment_l_score>0 and divide(p4_1y_comment_h_score,p4_1y_comment_l_score)>3;
![错误sql](https://user-images.githubusercontent.com/35916131/119758447-230a7280-bed9-11eb-89a8-7cb88e4e2bb1.jpg)

But when use p4_1y_comment_l_score to projection, it is can query normally.
sql2:
select p4_1y_comment_l_score  from xxxx where p4_1y_comment_l_score>0 and divide(p4_1y_comment_h_score,p4_1y_comment_l_score)>3;
![正常sql](https://user-images.githubusercontent.com/35916131/119758471-2dc50780-bed9-11eb-81ef-e85d60b08d96.jpg)

The stack（Forgive me can just privide the stack in picture ,because intranet in company）：
![错误栈1](https://user-images.githubusercontent.com/35916131/119758485-361d4280-bed9-11eb-9965-4e75e9ee3820.jpg)
![错误栈2](https://user-images.githubusercontent.com/35916131/119758542-4503f500-bed9-11eb-98cf-7196357d06fb.jpg)
![错误栈2](https://user-images.githubusercontent.com/35916131/119758569-4e8d5d00-bed9-11eb-9927-bc55354103b1.jpg)

TraceLog：
sql1：
select customer_no from xxxx where p4_1y_comment_l_score>0 and divide(p4_1y_comment_h_score,p4_1y_comment_l_score)>3;
![错误TraceLog](https://user-images.githubusercontent.com/35916131/119760155-13d8f400-bedc-11eb-88a7-1903974250af.jpg)

sql2:
select p4_1y_comment_l_score  from xxxx where p4_1y_comment_l_score>0 and divide(p4_1y_comment_h_score,p4_1y_comment_l_score)>3;
![正确TraceLog](https://user-images.githubusercontent.com/35916131/119760167-1a676b80-bedc-11eb-8974-82e94ac8f501.jpg)

**Comment by 5herhom** [comment]
Although I use sub query , the issue is also accured.
![子句](https://user-images.githubusercontent.com/35916131/119771350-22c9a180-bef0-11eb-923b-6c2c0ae23ebc.jpg)

**Comment by amosbird** [comment]
Heh, not the "projection" I'm expecting :)

**Comment by vdimir** [comment]
Subquery don't solve issue because of optimization and condition for division result moved to subquery automatically:
```
create table foo (val Decimal(32,6), v2 Decimal(32,6)) engine = Memory;
insert into foo values (1,2),(4,2),(3,0);
```

```
EXPLAIN SYNTAX
SELECT *
FROM
(
    SELECT *
    FROM foo
    WHERE v2 > 0
)
WHERE (val / v2) > 2
```

```
┌─explain─────────────────────────────────┐
│ SELECT                                  │
│     val,                                │
│     v2                                  │
│ FROM                                    │
│ (                                       │
│     SELECT                              │
│         val,                            │
│         v2                              │
│     FROM foo                            │
│     WHERE (v2 > 0) AND ((val / v2) > 2) │
│ )                                       │
│ WHERE (val / v2) > 2                    │
└─────────────────────────────────────────┘
```

Workaround is to disable `enable_optimize_predicate_expression`, but it's not best solution.

```
SELECT * FROM ( SELECT * FROM foo WHERE v2 > 0 ) WHERE (val / v2) >= 2 SETTINGS enable_optimize_predicate_expression = 0
```

Maybe you can change expression from `p4_1y_comment_h_score / p4_1y_comment_l_score > 3` to `p4_1y_comment_h_score > 3 * p4_1y_comment_l_score` if it won't overflow.

It will work with `if` or `and` when short circuit function evaluation will be implemented, but it's WIP now https://github.com/ClickHouse/ClickHouse/issues/12587

**Comment by 5herhom** [comment]
> Subquery don't solve issue because of optimization and condition for division result moved to subquery automatically:
> 
> ```
> create table foo (val Decimal(32,6), v2 Decimal(32,6)) engine = Memory;
> insert into foo values (1,2),(4,2),(3,0);
> ```
> 
>  ```
> EXPLAIN SYNTAX
> SELECT *
> FROM
> (
>     SELECT *
>     FROM foo
>     WHERE v2 > 0
> )
> WHERE (val / v2) > 2
> ```
> 
>  ```
> ┌─explain─────────────────────────────────┐
> │ SELECT                                  │
> │     val,                                │
> │     v2                                  │
> │ FROM                                    │
> │ (                                       │
> │     SELECT                              │
> │         val,                            │
> │         v2                              │
> │     FROM foo                            │
> │     WHERE (v2 > 0) AND ((val / v2) > 2) │
> │ )                                       │
> │ WHERE (val / v2) > 2                    │
> └─────────────────────────────────────────┘
> ```
> 
> Workaround is to disable `enable_optimize_predicate_expression`, but it's not best solution.
> 
> ```
> SELECT * FROM ( SELECT * FROM foo WHERE v2 > 0 ) WHERE (val / v2) >= 2 SETTINGS enable_optimize_predicate_expression = 0
> ```
> 
> Maybe you can change expression from `p4_1y_comment_h_score / p4_1y_comment_l_score > 3` to `p4_1y_comment_h_score > 3 * p4_1y_comment_l_score` if it won't overflow.
> 
> It will work with `if` or `and` when short circuit function evaluation will be implemented, but it's WIP now #12587

Thank you for your advice.
Another question is that  when execute query below, the "where" condition should return false  after execute "v2 > 0" for the v2=0, but it still execute "(val / v2) > 2" which causes zero divide exception.   In my understanding, it is unnormal.

> SELECT                              
>          val,           
>          v2             
>      FROM foo       
>      WHERE (v2 > 0) AND ((val / v2) > 2)

**Comment by vdimir** [comment]
>  In my understanding, it is unnormal.

It's because lazy evaluation for columns not implemented yet https://github.com/ClickHouse/ClickHouse/issues/12587

**Comment by 5herhom** [comment]
> > In my understanding, it is unnormal.
> z
> It's because lazy evaluation for columns not implemented yet #12587

**Comment by 5herhom** [comment]
> > In my understanding, it is unnormal.
> 
> It's because lazy evaluation for columns not implemented yet #12587
If it's lazy evaluation, why it throw exception when show column custimer_no, but query whitout exception where show column p3_service_last_2y_use_cnt. It's just different in the col to show... 
![错误和正确对比jpg](https://user-images.githubusercontent.com/35916131/120128531-d08cc700-c1f4-11eb-8410-fd13e205d569.jpg)


And they are not use to be primary key , order by key or shard key.
![建表](https://user-images.githubusercontent.com/35916131/120128544-d71b3e80-c1f4-11eb-978b-ef13b91b241f.jpg)
![分布式表建表](https://user-images.githubusercontent.com/35916131/120128798-72141880-c1f5-11eb-8bb6-8dfd511e3ae0.png)

**Comment by Avogar** [comment]
After merging https://github.com/ClickHouse/ClickHouse/pull/23367 such query works:
```sql
SELECT                              
    val,           
    v2             
FROM foo       
WHERE (v2 > 0) AND ((val / v2) > 2)
```
